design/one-pager-managed-resource-api-design.md

```diff
@@ -115,13 +119,26 @@ Related to https://github.com/crossplaneio/crossplane/issues/624
 The decision for an external resource name is made by the controller of that managed resource in `Create` phase.
 Possible cases are as following:
 * The provider doesn't allow us to specify a name. Fetch it after its client's `POST` call. Override `crossplane.io/external-name`
-annotation's value with what you get as name.
+  annotation's value with what you get as name.
 * The provider allows to specify a name.
   * If `crossplane.io/external-name` annotation has a value use it. If it's empty;
   * Use managed resource's name and write that into `crossplane.io/external-name` annotation.
 
 Use the value of that annotation as external resource name in _all_ queries.
 
+#### External Resource Labeling
+
+If the external resource supports labelling, we should label it with the managed
+resource name. While this is helpful for services with non-deterministic naming,
+it'd benefit the services that allow you to name the external resource, too,
+since the labels could be used for searching and batch operations on the cloud level.
+The key to use in those labels is `crossplane.io/resource-name` and the value is
+the name of the managed resource. An example would look like:
+```
+A tag for a VPC in AWS:
+  "crossplane.io/resource-name": "myappnamespace-mynetwork-5sc8a"
```

Comment:
@negz @hasheddan I am actually not sure about the `resource-name` part here. Should it be the kind of the managed resource? Like `crossplane.io/cloudsqlinstance` or `crossplane.io/cloudsqlinstance-name`?

Reply:
I think either way would be fine, but I agree that using the actual resource type in the tag would make it easier to identify.

Reply:
If we wanted to encode the managed resource kind (as Crossplane sees it), perhaps we should do that with a separate label? I imagine the relationship would be pretty obvious though (i.e. a `cloudsqlinstance` is never not going to be a CloudSQL instance in GCP and vice versa).

Reply:
> I imagine the relationship would be pretty obvious though (i.e. a cloudsqlinstance is never not going to be a CloudSQL instance in GCP and vice versa).

I agree. But I'm thinking more like how a user sees an external resource and be able to find it in the cluster easily and exact kind string could help. User knows that the service is Cloud SQL or Azure Database but the kind name is a bit different in some cases, for example https://github.com/crossplaneio/stack-azure/issues/65

How about the following pair?
```
"crossplane.io/resource-name": "myappnamespace-mynetwork-5sc8a"
"crossplane.io/resource-kind": "cloudsqlinstance"
```
or
```
"crossplane.io/name": "myappnamespace-mynetwork-5sc8a"
"crossplane.io/kind": "cloudsqlinstance"
```
Maybe `resource` is unnecessary?

Reply:
@hasheddan @negz ping

Reply:
@muvaf apologies for letting this slip by, I would advocate for the first of the two options you propose here :)

Reply:
> I'm thinking more like how a user sees an external resource and be able to find it in the cluster easily and exact kind string could help.

Ah - it took me a few minutes just now to grasp this concept but it makes sense to me now. I see the value in adding the kind as a label/tag and think we should do it.

> I would advocate for the first of the two options you propose here.

What's your thinking here @hasheddan? My suspicion is that we're going to run into a lot of weird and wonderful limitations across different providers around what characters and lengths are valid for tags and labels, so we should probably pick the shortest, simplest keys possible.

Reply:
@negz my primary thinking was that `resource-*` was more explicit, but I could go either way. On this topic, what is our course of action if whatever structure we choose is not acceptable to the provider?

Reply:
Got it! I'd prefer to leave off the `resource-` prefix. Ideally it's implied (I don't expect us to be needing to label anything in the provider that did not correspond to a Crossplane managed resource).

> What is our course of action if whatever structure we choose is not acceptable to the provider?

I think we're going to have to be pretty flexible around how this is implemented in crossplane-runtime. For example I imagine in some tagging systems we're going to have to add single tags like `crossplane.io/name=foo`, in other systems we're going to need to live without the `./` characters, in some we're going to have to live within character limits (`c8e.io/name`?). Given these labels are more for humans than automation I think it's okay to be a little inconsistent if we need to, as long as we can come up with a format that communicates "this is a Crossplane thing" in each case.

Reply:
@negz what would you think about making a `managed.WithTagger` parameter for the managed reconciler that used the pattern @muvaf outlines above by default but also allows for overriding? Could potentially allow for a stack to define a single tagging strategy that could be used across all of their resource types, which would help enforce consistency across a given provider.

Reply:
@hasheddan I think what we'll need from managed resource developers is a function that takes at least `key string, value string` parameters because Crossplane doesn't know which field of the resource is used for tagging, this part could be left to the developer. But I feel like we'll discover the edge cases when we do start to implement this. As you mentioned, they should be able to override the default, too. Maybe we'd just leave it completely to the developer.